File: ruby/leet_code/85.rb
# Module: LeetCode
# This module contains solutions for problems from LeetCode.
module LeetCode
  # Function: maximal_rectangle
  # This function calculates the maximal rectangle in a matrix.
  #
  # @param matrix [Array<Array<String>>] The input matrix.
  # @return [Integer] The area of the maximal rectangle.
  def maximal_rectangle(matrix)
    return 0 if matrix.empty?

    m, n = matrix.size, matrix[0].size
    left, right, height = Array.new(n, 0), Array.new(n, n), Array.new(n, 0)
    max_area = 0

    (0...m).each do |i|
      cur_left, cur_right = 0, n

      (0...n).each do |j|
        if matrix[i][j] == '1'
          height[j] += 1
          left[j] = [left[j], cur_left].max
        else
          height[j] = 0
          left[j] = 0
          cur_left = j + 1
        end
      end

      (n - 1).downto(0) do |j|
        if matrix[i][j] == '1'
          right[j] = [right[j], cur_right].min
        else
          right[j] = n
          cur_right = j
        end
      end

      max_area = [max_area, *(0...n).map { |j| (right[j] - left[j]) * height[j] }].max
    end

    max_area
  end
end

File: typescript/leet_code/85.js
/**
 * @fileoverview This module provides a function to calculate the maximal rectangle in a binary matrix.
 * @author jerodg
 */
/**
 * Calculates the maximal rectangle in a binary matrix.
 * The function uses dynamic programming to calculate the maximum height, left, and right boundary for each cell in the matrix.
 * The maximum area of the rectangle is then calculated as the height times the difference between the right and left boundaries.
 * This is done for each cell in the matrix, and the maximum area is updated if a larger area is found.
 * @param {string[][]} matrix - The input binary matrix.
 * @return {number} The area of the maximal rectangle.
 */
function maximalRectangle(matrix) {
    // Early return if the matrix is empty
    if (matrix.length === 0 || matrix[0].length === 0)
        return 0;
    var rows = matrix.length;
    var cols = matrix[0].length;
    var maxArea = 0;
    // Initialize arrays to store the maximum height and the left and right boundaries for each cell
    var left = new Array(cols).fill(0);
    var right = new Array(cols).fill(cols);
    var height = new Array(cols).fill(0);
    // Iterate over each cell in the matrix
    for (var i = 0; i < rows; i++) {
        var curLeft = 0, curRight = cols;
        for (var j = 0; j < cols; j++) {
            // If the cell is '1', increment the height and update the left boundary
            if (matrix[i][j] === "1") {
                height[j]++;
                left[j] = Math.max(left[j], curLeft);
            }
            else {
                // If the cell is '0', reset the height and left boundary, and update curLeft for the next cell
                height[j] = 0;
                left[j] = 0;
                curLeft = j + 1;
            }
        }
        // Iterate from right to left to update the right boundaries
        for (var j = cols - 1; j >= 0; j--) {
            if (matrix[i][j] === "1") {
                right[j] = Math.min(right[j], curRight);
            }
            else {
                right[j] = cols;
                curRight = j;
            }
        }
        // Calculate the maximum area for each cell
        for (var j = 0; j < cols; j++) {
            maxArea = Math.max(maxArea, (right[j] - left[j]) * height[j]);
        }
    }
    // Return the maximum area
    return maxArea;
}
